contracts/contracts/TACoApplication.sol

```diff
@@ -957,4 +1051,39 @@ contract TACoApplication is
         stakingProviderWrapper[0] = _stakingProvider;
         tStaking.seize(_penalty, 100, _investigator, stakingProviderWrapper);
     }
+
+    /**
+     * @notice Penalize the staking provider's future reward
+     * @param _stakingProvider Staking provider address
+     */
+    function penalize(address _stakingProvider) external updateReward(_stakingProvider) {
+        require(
+            msg.sender == address(childApplication),
+            "Only child application allowed to penalize"
+        );
+
+        if (_stakingProvider == address(0)) {
+            return;
+        }
+
+        StakingProviderInfo storage info = stakingProviderInfo[_stakingProvider];
+        uint96 before = effectiveAuthorized(info.authorized, info.penaltyPercent);
+        info.endPenalty = uint64(block.timestamp + penaltyDuration);
+        info.penaltyPercent = penaltyDefault;
+        if (info.operatorConfirmed) {
+            authorizedOverall -= before - effectiveAuthorized(info.authorized, info.penaltyPercent);
+        }
+        emit Penalized(_stakingProvider, info.penaltyPercent, info.endPenalty);
+    }
+
+    /**
+     * @notice Resets future reward back to 100%
+     * @param _stakingProvider Staking provider address
+     */
+    function resetReward(address _stakingProvider) external {
```

Comment:
Can you clarify what happens in `resetReward` is not called for a penalized staking provider whose penalty ended?

Reply:
`resetReward` also included in `updareReward`, so if it none such methods will be called then reward calculation will be looking like staker still has penalty. I know it's not great but with our model - we need active input to update values

Reply:
✔️